be/src/exec/vectorized/hudi_mor_scanner.cpp

```diff
@@ -0,0 +1,348 @@
+    _scanner_get_next_chunk = _jni_env->GetMethodID(_hudi_mor_scanner_cls, "nextChunkOffHeap", "()J");
+    DCHECK(_scanner_get_next_chunk != nullptr);
+    _scanner_column_release = _jni_env->GetMethodID(_hudi_mor_scanner_cls, "releaseOffHeapColumnVector", "(I)V");
+    DCHECK(_scanner_column_release != nullptr);
+
+    _jni_env->DeleteLocalRef(jbase_path);
+    _jni_env->DeleteLocalRef(jhive_column_names);
+    _jni_env->DeleteLocalRef(jhive_column_types);
+    _jni_env->DeleteLocalRef(jrequired_fields);
+    _jni_env->DeleteLocalRef(jinstant_time);
+    _jni_env->DeleteLocalRef(jlog_paths);
+    _jni_env->DeleteLocalRef(jdata_file_path);
+    _jni_env->DeleteLocalRef(jserde);
+    _jni_env->DeleteLocalRef(jinstant_time);
+    _jni_env->DeleteLocalRef(jstring_class);
+    _jni_env->DeleteLocalRef(hudi_mor_scanner_util_cls);
+    return Status::OK();
+}
+
+Status HudiMORScanner::_has_next(bool* result) {
+    jboolean ret = _jni_env->CallBooleanMethod(_hudi_mor_scanner, _scanner_has_next);
+    CHECK_JAVA_EXCEPTION("call hudi MOR scanner hasNext failed")
+    *result = ret;
+    return Status::OK();
+}
+
+Status HudiMORScanner::_get_next_chunk(long* chunk_meta) {
+    SCOPED_TIMER(_profile.io_timer);
+    COUNTER_UPDATE(_profile.io_counter, 1);
+    *chunk_meta = _jni_env->CallLongMethod(_hudi_mor_scanner, _scanner_get_next_chunk);
+    CHECK_JAVA_EXCEPTION("getNextChunk failed")
+    return Status::OK();
+}
+
+template <PrimitiveType type, typename CppType>
+void HudiMORScanner::_append_data(Column* column, CppType& value) {
+    auto appender = [](auto* column, CppType& value) {
+        using ColumnType = typename vectorized::RunTimeColumnType<type>;
+        ColumnType* runtime_column = down_cast<ColumnType*>(column);
+        runtime_column->append(value);
+    };
+
+    if (column->is_nullable()) {
+        auto* nullable_column = down_cast<NullableColumn*>(column);
+        auto* data_column = nullable_column->data_column().get();
+        NullData& null_data = nullable_column->null_column_data();
+        null_data.push_back(0);
+        appender(data_column, value);
+    } else {
+        appender(column, value);
+    }
+}
+
+Status HudiMORScanner::_process_null_value(Column* column, SlotDescriptor* slot_desc) {
+    if (!column->is_nullable()) {
+        return Status::DataQualityError(
+                fmt::format("Unexpected NULL value occurs on NOT NULL column[{}]", slot_desc->col_name()));
+    }
+    column->append_nulls(1);
+    return Status::OK();
+}
+
+template <PrimitiveType type, typename CppType>
+Status HudiMORScanner::_append_primitive_data(long num_rows, long* chunk_meta_ptr, int& chunk_meta_index,
+                                              ColumnPtr& column, SlotDescriptor* slot_desc) {
+    bool* null_column_ptr = reinterpret_cast<bool*>(chunk_meta_ptr[chunk_meta_index++]);
+    CppType* column_ptr = reinterpret_cast<CppType*>(chunk_meta_ptr[chunk_meta_index++]);
+    for (int i = 0; i < num_rows; i++) {
```

Comment:
Since the memory layout is identical to native one, you can `memcpy` the entire buffer instead of copy one by one